homeassistant/components/sensor/buienradar.py

```diff
@@ -0,0 +1,471 @@
+                # no section with weather stations
+                _LOGGER.warning("Missing section in Buienradar xmldata (%s)."
+                                "Can happen 00:00-01:00 CE(S)T",
+                                CONF_BR_WEERSTATIONS)
+                return None
+
+            xmldata_tmp = xmldata_tmp[CONF_BR_WEERSTATIONS]
+            if CONF_BR_WEERSTATION not in xmldata_tmp:
+                # no weatherstation section(s)
+                _LOGGER.warning("Missing section in Buienradar xmldata (%s)."
+                                "Can happen 00:00-01:00 CE(S)T",
+                                CONF_BR_WEERSTATION)
+                return None
+
+            xmldata_tmp = xmldata_tmp[CONF_BR_WEERSTATION]
+            for wstation in xmldata_tmp:
+                wslat = float(wstation[CONF_BR_LAT])
+                wslon = float(wstation[CONF_BR_LON])
+
+                if ((loc_data is None) or (
+                        distance(self.coordinates[CONF_LATITUDE],
+                                 self.coordinates[CONF_LONGITUDE],
+                                 wslat,
+                                 wslon) < dist)):
+                    dist = distance(self.coordinates[CONF_LATITUDE],
+                                    self.coordinates[CONF_LONGITUDE],
+                                    wslat, wslon)
+                    self.weatherstation = wstation[CONF_BR_STATIONCODE]
+                    loc_data = wstation
+
+            if loc_data is None:
+                _LOGGER.warning("No weatherstation selected; aborting update.")
+                return None
+            else:
+                _LOGGER.debug("Selected station: code='%s', "
+                              "name='%s', lat='%s', lon='%s'.",
+                              loc_data[CONF_BR_STATIONCODE],
+                              loc_data[CONF_BR_STATIONNAME][CONF_BR_TEXT],
+                              loc_data[CONF_BR_LAT],
+                              loc_data[CONF_BR_LON])
+                return loc_data
+
+        def try_again(err: str):
+            """Retry in 2 minutes."""
+            _LOGGER.warning('Retrying in 2 minutes: %s', err)
+            nxt = dt_util.utcnow() + timedelta(minutes=2)
+            async_track_point_in_utc_time(self.hass, self.async_update,
+                                          nxt)
+
+        def local_time_offset(convt=None):
+            """Return offset of local zone from UTC."""
+            # python2.3 localtime() can't take None
+            if convt is None:
+                convt = tm.time()
+
+            if tm.localtime(convt).tm_isdst and tm.daylight:
+                _LOGGER.debug('local_time_offset: %s', -tm.altzone)
+                return -tm.altzone
+            else:
+                _LOGGER.debug('local_time_offset: %s', -tm.timezone)
+                return -tm.timezone
+
+        # get weather data
+        resp = None
+        try:
+            websession = async_get_clientsession(self.hass)
```

Comment:
There is way too much going on in 1 method. Fetching, parsing, processing, collecting. Please extract the logic into an external lib that can be included. The external lib could be responsible for just parsing/processing/collecting and you can leave fetching in Hass.